onnxruntime/core/optimizer/qdq_transformer/qdq_selectors.cc

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#if !defined(ORT_MINIMAL_BUILD)
+
+#include "core/optimizer/qdq_transformer/qdq_selectors.h"
+
+#include "core/graph/graph.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/qdq_transformer/qdq_util.h"
+#include "core/optimizer/utils.h"
+
+namespace onnxruntime {
+namespace QDQ {
+namespace {
+// adjust for an optional input that has an entry but does not exist
+int NumActualInputs(const Node& node) {
+  const auto& input_defs = node.InputDefs();
+  return gsl::narrow_cast<int>(std::count_if(input_defs.cbegin(), input_defs.cend(),
+                                             [](const NodeArg* def) { return def && def->Exists(); }));
+}
+}  // namespace
+
+bool BaseSelector::CheckQDQNodes(const Graph& graph, const Node& node,
+                                 const std::vector<const Node*>& dq_nodes,
+                                 const std::vector<const Node*>& q_nodes,
+                                 int num_dq_inputs) const {
+  if (num_dq_inputs == -1) {
+    num_dq_inputs = NumActualInputs(node);
+  }
+
+  return num_dq_inputs == gsl::narrow_cast<int>(dq_nodes.size()) &&
+         node.OutputDefs().size() == q_nodes.size() &&
+         optimizer_utils::CheckOutputEdges(graph, node, q_nodes.size());
```

Comment:
>optimizer_utils::CheckOutputEdges(graph, node, q_nodes.size()) [](http://example.com/codeflow?start=9&length=62)

perhaps I'm missing something - does this check also cover the check in the previous line?

Reply:
Not quite as it just checks the edges and not what type of node they're connected to.

It may however be better to check that the number of outputs which exist (handle optional output that is in OutputDefs() but isn't present) matches the number of q_nodes (so 1:1 between an output and a Q), and that this node doesn't produce a graph output instead of using CheckOutputEdges.